### PR TITLE
add homa-validator-list module

### PR DESCRIPTION
**Pull request by wangjj9219**

### CODE REVIEW

File: modules/homa-validator-list/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+			#[pallet::compact] amount: Balance,
+		) -> DispatchResultWithPostInfo {
+			let guarantor = ensure_signed(origin)?;
+			let free_balance = T::LiquidTokenCurrency::free_balance(&guarantor);
+			let total_should_locked = Self::total_locked_by_guarantor(&guarantor).unwrap_or_default();
+
+			if let Some(extra) = free_balance.checked_sub(total_should_locked) {
+				let amount = amount.min(extra);
+
+				if amount.is_zero() {
```
Comment (xlc) [inline modules/homa-validator-list/src/lib.rs:207]:
```suggestion
				if !amount.is_zero() {
```

File: modules/homa-validator-list/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+		#[pallet::weight(T::WeightInfo::unbond())]
+		#[transactional]
+		pub fn unbond(
+			origin: OriginFor<T>,
+			validator: T::RelaychainAccountId,
+			#[pallet::compact] amount: Balance,
+		) -> DispatchResultWithPostInfo {
+			let guarantor = ensure_signed(origin)?;
+
+			if amount.is_zero() {
```
Comment (xlc) [inline modules/homa-validator-list/src/lib.rs:232]:
```suggestion
			if !amount.is_zero() {
```

File: modules/homa-validator-list/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+		#[pallet::weight(T::WeightInfo::rebond())]
+		#[transactional]
+		pub fn rebond(
+			origin: OriginFor<T>,
+			validator: T::RelaychainAccountId,
+			#[pallet::compact] amount: Balance,
+		) -> DispatchResultWithPostInfo {
+			let guarantor = ensure_signed(origin)?;
+
+			if amount.is_zero() {
```
Comment (xlc) [inline modules/homa-validator-list/src/lib.rs:260]:
```suggestion
			if !amount.is_zero() {
```

File: modules/homa-validator-list/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+				Ok(())
+			})?;
+			Ok(().into())
+		}
+
+		#[pallet::weight(T::WeightInfo::freeze())]
+		#[transactional]
+		pub fn freeze(origin: OriginFor<T>, validators: Vec<T::RelaychainAccountId>) -> DispatchResultWithPostInfo {
+			T::FreezeOrigin::ensure_origin(origin)?;
+			if !validators.len().is_zero() {
```
Comment (xlc) [inline modules/homa-validator-list/src/lib.rs:303]: no need to have check. for_each will just do nothing if vec is empty

File: modules/homa-validator-list/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+				});
+			}
+			Ok(().into())
+		}
+
+		#[pallet::weight(T::WeightInfo::thaw())]
+		#[transactional]
+		pub fn thaw(origin: OriginFor<T>, validators: Vec<T::RelaychainAccountId>) -> DispatchResultWithPostInfo {
+			T::SlashOrigin::ensure_origin(origin)?;
+			if !validators.len().is_zero() {
```
Comment (xlc) [inline modules/homa-validator-list/src/lib.rs:322]: no need to have this check

**Comment by wangjj9219** [comment]
1. Basically implement #757 , some difference with the original design：
   a. the shares and rewards can use incentive/rewards, need to do some modification for incentive/rewards.
   b. add `ValidatorInsuranceThreshold` to determine whether the validator is valid.
2. Split the process of `staking_pool::rebalance` into multiple phases.

**Review by xlc — APPROVED** [review]

**Pull request merged**